Earthfile

```diff
@@ -39,241 +63,273 @@ setup:
+  COPY scripts ./
+  COPY excluded-pkgs.txt spaceros-pkgs.txt spaceros.repos ./
 
-repos-file:
-  FROM +setup
-  COPY excluded-pkgs.txt ./
-  COPY spaceros-pkgs.txt ./
-  COPY spaceros.repos ./
   # This is a fresh image, so we do not need to exclude installed packages.
-  RUN --no-cache sh scripts/generate-repos.sh \
+  ENV AMENT_PREFIX_PATH=${SPACEROS_DIR}
+  RUN sh generate-repos.sh \
                  --outfile ros2.repos \
                  --packages spaceros-pkgs.txt \
                  --excluded-packages excluded-pkgs.txt \
-                 --exclude-installed false \
                  --rosdistro ${ROS_DISTRO}
-  RUN --no-cache python3 scripts/merge-repos.py ros2.repos spaceros.repos -o output.repos
+  RUN  python3 merge-repos.py ros2.repos spaceros.repos -o output.repos
   SAVE ARTIFACT output.repos AS LOCAL ros2.repos
 
-spaceros-artifacts:
-  # we must run it in a separate container, so that downstream tasks can be cached if vcs file does not change
-  FROM +setup
-  USER ${USERNAME}
-  WORKDIR ${SPACEROS_DIR}
-
-  COPY ros2.repos ./
-  COPY excluded-pkgs.txt ./
-
-  # this ensure the vcs import and export results are not cached
-  RUN --no-cache echo "Cloning spaceros repo artifacts"
-
-  # we run vcstool inside this task, because some packages in `ros2.repos` are not pinned and otherwise
-  # earthly won't pull latest changes
-  RUN mkdir src
-  RUN vcs import --shallow --retry 3 --input ros2.repos src
-  RUN vcs export --exact src > exact.repos
-
-  SAVE ARTIFACT ros2.repos
-  SAVE ARTIFACT exact.repos # `ros2.repos`, but with pinned versions (e.g. SHAs instead of branches)
-  SAVE ARTIFACT excluded-pkgs.txt
+###############################################################################
+### IKOS Installation
+# This stage is responsible for installing IKOS and its dependencies.
+# This will be used for static analysis of the ROS 2 workspace during build
+# test and development.
+###############################################################################
+ikos-install:
+  FROM +pre-installation
+
+  RUN apt-get install --yes \
+  gcc g++ cmake libgmp-dev libboost-dev libboost-filesystem-dev \
+  libboost-thread-dev libboost-test-dev \
+  libsqlite3-dev libtbb-dev libz-dev libedit-dev \
+  python3 python3-pip python3-venv \
+  llvm-14 llvm-14-dev llvm-14-tools clang-14 ros-dev-tools
 
+  RUN git clone -b v3.2 --depth 1 https://github.com/NASA-SW-VnV/ikos.git
+  RUN cd ikos \
+      && mkdir build \
+      && cd build \
+      && cmake \
+      -DCMAKE_INSTALL_PREFIX="/opt/ikos" \
+      -DCMAKE_BUILD_TYPE="Release" \
+      -DLLVM_CONFIG_EXECUTABLE="/usr/lib/llvm-14/bin/llvm-config" \
+      .. \
+      && make -j$(nproc) \
+      && make install
+  ENV PATH="/opt/ikos/bin/:$PATH"
+  RUN rm -rf ikos
+
+  SAVE ARTIFACT /opt/ikos AS LOCAL ikos
+
+ADD_IKOS:
+  FUNCTION
+    COPY +ikos-install/ikos /opt/ikos
+    ENV PATH="/opt/ikos/bin/:$PATH"
+
+    RUN apt-get install --yes \
```

Comment:
Since this is a function, I think it wouldn't make a difference. But to keep things properly structured, I agree with swapping the order. I will update it.

Reply:
After looking into it a bit, I realized this will save some more build time.